plugin/trino-kudu/src/main/java/io/trino/plugin/kudu/schema/SchemaEmulationByTableNameConvention.java

```diff
@@ -167,17 +151,29 @@ private void createAndFillSchemasTable(KuduClientWrapper client)
         Schema schema = new Schema(ImmutableList.of(schemaColumnSchema));
         CreateTableOptions options = new CreateTableOptions();
         options.addHashPartitions(ImmutableList.of(schemaColumnSchema.getName()), 2);
-        KuduTable schemasTable = client.createTable(rawSchemasTableName, schema, options);
-        KuduSession session = client.newSession();
-        try {
+
+        KuduTable schemasTable = createTableIfNotExists(client, schema, rawSchemasTableName, options);
+        try (KuduOperationApplier operationApplier = KuduOperationApplier.fromKuduClientWrapper(client)) {
             for (String schemaName : existingSchemaNames) {
-                Insert insert = schemasTable.newInsert();
-                insert.getRow().addString(0, schemaName);
-                applyOperationAndVerifySucceeded(session, insert);
+                Upsert upsert = schemasTable.newUpsert();
+                upsert.getRow().addString(0, schemaName);
+                operationApplier.applyOperationAsync(upsert);
             }
         }
-        finally {
-            session.close();
+    }
+
+    private static KuduTable createTableIfNotExists(KuduClientWrapper client, Schema schema, String name, CreateTableOptions options)
+            throws KuduException
+    {
+        try {
+            return client.createTable(name, schema, options);
+        }
+        catch (KuduException e) {
+            if (e.getStatus().isAlreadyPresent()) {
+                // Table already exists
+                return client.openTable(name);
```

Comment:
Q: (For complete correctness) should we do a comparison on the existing table's options to be sure they're effectively "the same"?
(I know this is fixing a race in test code, but in Prod code if someone attempted to create a table w/incompatible options at the same time as someone else, they may (unexpectedly) get back a KuduTable which doesn't match the create options they were trying to apply?)

Reply:
(Looking a little more closely, it seems the only usage is from `createAndFillSchemasTable`, which means the case I suggested can't manifest - so ignore!).

Reply:
The whole approach of emulating schemas in kudu using a kudu table is not safe given[ kudu's consistency semantics ](https://kudu.apache.org/docs/transaction_semantics.html) and trino does not propagate client timestamps in kudu.

I talked with some kudu maintainers, impala uses the hive metastore to emulate schemas in kudu, that introduces a big dependency however.